### PR TITLE
bugfix: take lmhlo_disc ops into consideration when doing fusion

### PR DESCRIPTION
**Pull request by wyzero**

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/fusion_utils.cc [hunk 1]
```diff
@@ -401,7 +402,15 @@ bool isFusible(Operation* op) {
 // For some ops (e.g. lmhlo ops), some operands are the output memrefs
 // Thus these operands are supposed to be updated.
 int getNumResultOperands(Operation* op) {
-  if (op->getDialect()->getTypeID() != TypeID::get<lmhlo::LmhloDialect>()) {
+  if (auto customOp = dyn_cast_or_null<lmhlo_disc::CustomCallOp>(op)) {
```
Comment (fortianyou) [inline tao_compiler/mlir/disc/transforms/fusion_utils.cc:405]: What is the scenario of `lmhlo_disc::CustomCallOp`

File: tao_compiler/mlir/disc/transforms/tests/cpu-only-lmhlo-fusion.mlir [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// RUN: disc-opt -pass-pipeline='func.func(disc-fusion{gpu-enabled=false fusion-strategy=base})' -split-input-file %s -o - | FileCheck %s
+
+// CHECK-LABEL: @custom_call_op
+// CHECK-SAME: (%[[ARG0:.*]]: memref<?x?xf32, "cpu">, %[[ARG1:.*]]: memref<?x?xf32, "cpu">, %[[ARG2:.*]]: memref<?x?xf32, "cpu">, %[[ARG3:.*]]: memref<?x?xf32, "cpu">) -> memref<?x?xf32, "cpu">
+func @custom_call_op(%arg0: memref<?x?xf32, "cpu">, %arg1: memref<?x?xf32, "cpu">,
+                     %arg2: memref<?x?xf32, "cpu">, %arg3: memref<?x?xf32, "cpu">) -> memref<?x?xf32, "cpu"> {
+  // CHECK-NOT: "lmhlo.fusion"
+  "lmhlo.abs"(%arg0, %arg1) : (memref<?x?xf32, "cpu">, memref<?x?xf32, "cpu">) -> ()
+  "lmhlo_disc.custom_call"(%arg1, %arg2) {call_target_name = "test", disc.device = "cpu", has_side_effect = false, operand_segment_sizes = dense<[1, 1]> : vector<2xi32>} : (memref<?x?xf32, "cpu">, memref<?x?xf32, "cpu">) -> ()
```
Comment (fortianyou) [inline tao_compiler/mlir/disc/transforms/tests/cpu-only-lmhlo-fusion.mlir:9]: Is this test case used to guard and watch if `custom_call` is merged into a fusion group?

File: tao_compiler/mlir/disc/transforms/fusion_utils.cc (hunk 1 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/fusion_utils.cc:405]: For example, random_uniform, topk, quantized_conv ...

File: tao_compiler/mlir/disc/transforms/tests/cpu-only-lmhlo-fusion.mlir (hunk 2 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/tests/cpu-only-lmhlo-fusion.mlir:9]: No. It's used to make sure `lmhlo.abs` and `lmhlo.add` not be fused. Without this PR, they will be fused into one kernel, which is a bug.

File: tao_compiler/mlir/disc/transforms/fusion_utils.cc [hunk 3]
```diff
@@ -401,7 +402,15 @@ bool isFusible(Operation* op) {
...
+    // TODO(disc): add a registration base mechanism to support custom call op
+    // with more than one results.
+    return 1;
+  }
+
+  if (op->getDialect()->getTypeID() != TypeID::get<lmhlo::LmhloDialect>() &&
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/fusion_utils.cc:411]: why `LmhloDialect` and `LmhloDiscDialect` has not output operands?

Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/fusion_utils.cc:411]: Note that here is "not equal" ^v^

**Pull request merged**